compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -533,6 +540,44 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 // Recurse with the new expected type.
                 self.check_pat_inner(pat, opt_path_res, adjust_mode, inner_ty, new_pat_info)
             }
+            // If `pin_ergonomics` is enabled, peel the `&pin` from the pinned reference type. See the
+            // examples in `tests/ui/async-await/pin-ergonomics/`.
+            _ if self.tcx.features().pin_ergonomics()
+                && let AdjustMode::Peel { kind: peel_kind } = adjust_mode
+                && pat.default_binding_modes
+                && self.should_peel_smart_pointer(peel_kind, expected)
+                && let Some(pinned_ty) = expected.pinned_ty()
+                // Currently, only pinned reference is specially handled, leaving other
+                // pinned types (e.g. `Pin<Box<T>>` to deref patterns) handled as a
+                // deref pattern.
+                && let &ty::Ref(_, inner_ty, inner_mutability) = pinned_ty.kind() =>
+            {
+                debug!("scrutinee ty {expected:?} is a pinned reference, inserting pin deref");
+                // Preserve the pinned type. We'll need it later during THIR lowering.
+                self.typeck_results
+                    .borrow_mut()
+                    .pat_adjustments_mut()
+                    .entry(pat.hir_id)
+                    .or_default()
+                    .push(PatAdjustment { kind: PatAdjust::PinDeref, source: expected });
+
+                let binding_mode = adjust_binding_mode(Pinnedness::Pinned, inner_mutability);
+                // If the pinnedness is `Not`, it means the pattern is unpinned
+                // and thus requires an `Unpin` bound.
+                if binding_mode == ByRef::Yes(Pinnedness::Not, Mutability::Mut) {
+                    self.register_bound(
+                        inner_ty,
+                        self.tcx.require_lang_item(hir::LangItem::Unpin, Some(pat.span)),
+                        self.misc(pat.span),
+                    );
+                }
+                // Use the old pat info to keep `current_depth` to its old value.
+                let new_pat_info = PatInfo { binding_mode, ..old_pat_info };
+                // Recurse with the new expected type.
+                // using `break` instead of `return` in case where any shared codes are added
+                // after the `match pat.kind {}`.
```

Comment:
What does this mean?

Reply:
Oh, it's an obsolete comment that I forgot to remove.